File: javascript/leetcode/141. Linked List Cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
  if (!head) return false;
  let fast = head.next;

  while (fast) {
    if (fast === head) return true;
    fast = fast.next ? fast.next.next : null;
    head = head.next;
  }

  return false;
};

File: c/leetcode/141. Linked List Cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head)
{
  if (!head)
    return false;
  struct ListNode *fast = head->next;

  while (fast)
  {
    if (fast == head)
      return true;
    fast = fast->next ? fast->next->next : NULL;
    head = head->next;
  }

  return false;
}
